src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,231 @@
+end
+
+function BeliefPropagationCache(tn::ITensorNetwork, partitioned_vertices; kwargs...)
+  ptn = PartitionedGraph(tn, partitioned_vertices)
+  return BeliefPropagationCache(ptn; kwargs...)
+end
+
+function partitioned_itensornetwork(bp_cache::BeliefPropagationCache)
+  return bp_cache.partitioned_itensornetwork
+end
+messages(bp_cache::BeliefPropagationCache) = bp_cache.messages
+default_message(bp_cache::BeliefPropagationCache) = bp_cache.default_message
+function tensornetwork(bp_cache::BeliefPropagationCache)
+  return unpartitioned_graph(partitioned_itensornetwork(bp_cache))
+end
+
+#Forward from partitioned graph
+for f in [
+  :(NamedGraphs.partitioned_graph),
+  :(NamedGraphs.partitionedge),
+  :(NamedGraphs.partitionvertices),
+  :(NamedGraphs.vertices),
+  :(NamedGraphs.boundary_partitionedges),
+  :linkinds,
+]
+  @eval begin
+    function $f(bp_cache::BeliefPropagationCache, args...; kwargs...)
+      return $f(partitioned_itensornetwork(bp_cache), args...; kwargs...)
+    end
+  end
+end
+
+function default_message(bp_cache::BeliefPropagationCache, edge::PartitionEdge)
+  return default_message(bp_cache)(linkinds(bp_cache, edge))
+end
+
+function message(bp_cache::BeliefPropagationCache, edge::PartitionEdge)
+  mts = messages(bp_cache)
+  return get(mts, edge, default_message(bp_cache, edge))
+end
+function messages(
+  bp_cache::BeliefPropagationCache, edges::Vector{<:PartitionEdge}; kwargs...
+)
+  return [message(bp_cache, edge; kwargs...) for edge in edges]
+end
+
+function copy(bp_cache::BeliefPropagationCache)
+  return BeliefPropagationCache(
+    copy(partitioned_itensornetwork(bp_cache)),
+    copy(messages(bp_cache)),
+    default_message(bp_cache),
+  )
+end
+
+function default_bp_maxiter(bp_cache::BeliefPropagationCache)
+  return default_bp_maxiter(partitioned_graph(bp_cache))
+end
+function default_edge_sequence(bp_cache::BeliefPropagationCache)
+  return default_edge_sequence(partitioned_itensornetwork(bp_cache))
+end
+
+function incoming_messages(
+  bp_cache::BeliefPropagationCache,
+  partition_vertices::Vector{<:PartitionVertex};
+  ignore_edges=PartitionEdge[],
+)
+  bpes = boundary_partitionedges(bp_cache, partition_vertices; dir=:in)
+  ms = messages(bp_cache, setdiff(bpes, ignore_edges))
+  return reduce(vcat, ms; init=[])
+end
+
+function incoming_messages(
+  bp_cache::BeliefPropagationCache, partition_vertex::PartitionVertex; kwargs...
+)
+  return incoming_messages(bp_cache, [partition_vertex]; kwargs...)
+end
+
+function incoming_messages(bp_cache::BeliefPropagationCache, verts::Vector)
+  partition_verts = partitionvertices(bp_cache, verts)
+  messages = incoming_messages(bp_cache, partition_verts)
+  central_tensors = ITensor[
+    tensornetwork(bp_cache)[v] for v in setdiff(vertices(bp_cache, partition_verts), verts)
+  ]
+  return vcat(messages, central_tensors)
+end
+
+function incoming_messages(bp_cache::BeliefPropagationCache, vertex)
+  return incoming_messages(bp_cache, [vertex])
+end
```

Comment:
This can't distinguish between the case where the vertex name itself is a `Vector`, or if that refers to a collection of vertices.

To fix that, I think we could just get rid of the version `incoming_messages(bp_cache::BeliefPropagationCache, vertex)` and require that it must be a collection of vertices.

Reply:
Good point, I will get ride of that version.